### PR TITLE
Flax: avoid recompilation when params change

### LINKED ISSUE #1015: JAX / Flax: avoid recompilation on params change: guidance scale, size, number of steps
This is required to improve the demo(s), we currently have the sliders disabled.

Some users are also requesting it in discord and the forums. Our demo notebook is very fast on TPU hardware but users can't quickly explore different parameters.

### PR DESCRIPTION
**Pull request by pcuenca**
Fixes #1015

### CODE REVIEW

**Comment by camenduru** [comment]
Hi @pcuenca, I have a question 

in this doc https://jax.readthedocs.io/en/latest/_autosummary/jax.pmap.html

> static_broadcasted_argnums Calling the pmapped function with different values for these constants will trigger recompilation.

and we are sending prompt_ids: jnp.array to cores without saying static 
https://github.com/huggingface/diffusers/blob/b4a1ed85440d4d9c1cafbe118ca6c034000c85f9/src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py#L320

Is this means if we convert num_inference_steps, height, width, guidance_scale to jnp.array, can we send them like sending prompt_ids: jnp.array without recompilation 🙄

**Comment by pcuenca** [comment]
> Is this means if we convert num_inference_steps, height, width, guidance_scale to jnp.array, can we send them like sending prompt_ids: jnp.array without recompilation 🙄

Hi @camenduru, that's essentially right! However, there's a caveat: Jax will recompile the function if the tensor sizes change. So it will work for `guidance_scale`, but not for `width` or `height` because they will create a different `latents` tensor if they change. There are some tricks that essentially mean caching various versions of the compiled function, see for example here: https://github.com/google/jax/issues/2521.

**Comment by pcuenca** [comment]
This in-progress PR is meant to support the Stable Diffusion demo by enabling the advanced settings again. `guidance_scale` should be easy to achieve (it currently doesn't cause a recompilation, but we need to make it an array so we can batch queries using different scales).

However, `width` and `height` are problematic because they require different `latents` each time. We could create various combinations of static values and compile them separately. This is doable, but it makes batching more difficult because all the requests in a batch should have the same `(width, height)`. A couple of questions:
- Should we start with just `guidance_scale` for now? /cc @apolinario 
- @apolinario can you recommend a list of discretized pairs (width, height) that make sense?
- @patil-suraj do you have any recommendations to support various sizes that I might be missing?

Also, I'm ignoring `num_inference_steps` because I think is the least important of them all and it was added for performance considerations only.

**Comment by patil-suraj** [comment]
Thanks for the PR @pcuenca, think we can make `guidance_scale` an array to support batching. cc @patrickvonplaten 

Handling different sizes is always tricky in JAX.  So I would say let's first start with `guidance_scale` unless there's priority to handle various sizes. 

> @patil-suraj do you have any recommendations to support various sizes that I might be missing?

No, I used the same approach for BLOOM that you mentioned -> compile different versions of functions for different sizes. Maybe worth asking in the JAX chat.

**Comment by patrickvonplaten** [comment]
@pcuenca let me know when the PR is ready for review :-)

**Comment by dhruvrnaik** [comment]
Correct me if I am wrong, but `guidance_scale` should stay static if classifier free guidance condition is to be checked right?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 1]
```diff
@@ -260,13 +259,7 @@ def loop_body(step, args):
 
         # scale the initial noise by the standard deviation required by the scheduler
         latents = latents * self.scheduler.init_noise_sigma
-
-        if debug:
-            # run with python for loop
-            for i in range(num_inference_steps):
-                latents, scheduler_state = loop_body(i, (latents, scheduler_state))
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:267]: Think it's nice to have this option for easy debugging.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 2]
```diff
@@ -334,6 +326,12 @@ def __call__(
         height = height or self.unet.config.sample_size * self.vae_scale_factor
         width = width or self.unet.config.sample_size * self.vae_scale_factor
 
+        if isinstance(guidance_scale, float):
+            guidance_scale = jnp.array([guidance_scale] * prompt_ids.shape[0])
+            if len(prompt_ids.shape) > 2:
+                # Assume sharded
+                guidance_scale = guidance_scale.reshape(prompt_ids.shape[:2])
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:333]: Maybe put some comments here, explaining what this does.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 3]
```diff
@@ -389,7 +385,11 @@ def __call__(
 
 
 # TODO: maybe use a config dict instead of so many static argnums
-@partial(jax.pmap, static_broadcasted_argnums=(0, 4, 5, 6, 7, 9))
+@partial(
+    jax.pmap,
+    in_axes=(None, 0, 0, 0, None, None, None, 0, 0, 0),
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:390]: Very cool!

`in_axes` is 0 for non-static args, no ? Also, same comment as above, let's put some comments here.

**Review by patil-suraj — APPROVED** [review]
The PR looks good, just left a couple of comments ! Would also be nice if we could add a test for this. Thanks a lot for working on this!

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py (hunk 1 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:267]: The problem is that `debug` has to be passed from everywhere and it adds noise to the very busy `partial` here: https://github.com/huggingface/diffusers/blob/7db319bdae4023bbd6c7a7ca3d16abd873dcdbd0/src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py#LL388-L391

How do you feel about using a `DEBUG` constant instead, that we only use when we are really debugging?

Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:267]: I see, yeah setting DEBUG as a constant sounds good too!

**Pull request merged**